### PR TITLE
gpt-oss: checkpoint and rope

### PR DESCRIPTION
**Pull request by shuningjin**
# Description

1 Dequantization: reshape to be compatible with hf model loading bf16

2 Unscanned checkpoint

3 Yarn RoPE: 
- input complex representation: interleave -> allow concatenated
- truncate: True -> allow False
- attention scaling: none -> allow

See detail in b/437369407

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -618,6 +618,9 @@ rope_factor: 40
 beta_fast: 32
 beta_slow: 1
 mscale: 1.0
+rope_interleave: True # Complex representation is interleaved or concatenated
+rope_truncate: True # Round correction range
```
Comment (RissyRan) [inline MaxText/configs/base.yml]: shall we say "round a number to the nearest integer (floor/ceil)"

File: MaxText/configs/models/gpt-oss-20b.yml [hunk 2]
```diff
@@ -34,6 +34,9 @@ original_max_position_embeddings: 4096
 rope_factor: 32
 beta_fast: 32
 beta_slow: 1
+rope_interleave: False
+rope_truncate: False
+rope_attention_scaling: True
```
Comment (RissyRan) [inline MaxText/configs/models/gpt-oss-20b.yml]: let's also update DeepSeek v2, 3 for specific configs?

```
rope_interleave: True
rope_truncate: True
rope_attention_scaling: False
```

File: MaxText/convert_gpt_oss_unscanned_ckpt.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+}
+
+
+def _hf_to_maxtext_mapping(layer_idx: int = -1) -> dict:
+  """
+  Returns a mapping from HuggingFace model weight names to MaxText model weight names.
+
+  Args:
+    layer_idx (int): Layer index.
+    expert_idx (int): Expert index.
```
Comment (RissyRan) [inline MaxText/convert_gpt_oss_unscanned_ckpt.py]: Didn't have it here?

File: MaxText/convert_gpt_oss_unscanned_ckpt.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+      f"model.layers.{layer_idx}.mlp.router.weight": f"layers.{layer_idx}.feed_forward.gate.weight",
+      f"model.layers.{layer_idx}.mlp.router.bias": f"layers.{layer_idx}.feed_forward.gate.bias",
+      f"model.layers.{layer_idx}.mlp.experts.gate_up_proj": f"layers.{layer_idx}.feed_forward.experts.gate_up_proj",
+      f"model.layers.{layer_idx}.mlp.experts.gate_up_proj_bias": f"layers.{layer_idx}.feed_forward.experts.gate_up_proj_bias",
+      f"model.layers.{layer_idx}.mlp.experts.down_proj": f"layers.{layer_idx}.feed_forward.experts.down_proj",
+      f"model.layers.{layer_idx}.mlp.experts.down_proj_bias": f"layers.{layer_idx}.feed_forward.experts.down_proj_bias",
+  }
+
+
+def _convert_huggingface_to_jax_weights(base_model_path: str, model_size: str, model_params: dict, mem_info: psutil.Process):
```
Comment (RissyRan) [inline MaxText/convert_gpt_oss_unscanned_ckpt.py]: If you have the doc string already for above method. Let's align for others?

File: MaxText/layers/embeddings.py [hunk 5]
```diff
@@ -565,6 +571,9 @@ def __init__(
       rope_factor: float = 40,
       cast_as_fprop_dtype: bool = True,
       fprop_dtype: DType = jnp.bfloat16,
+      interleave=True,
+      truncate=True,
+      attention_scaling=False,
```
Comment (RissyRan) [inline MaxText/layers/embeddings.py]: Update doc string above?

File: MaxText/tests/forward_pass_logit_checker.py [hunk 6]
```diff
@@ -321,12 +321,16 @@ def main(config, test_args):  # pylint: disable=W0621
...
+        )
         assert jax.numpy.all(
             kl_div < test_args.max_kl_div,
         ), f"KL divergence values exceed the specified threshold of {test_args.max_kl_div}. Max divergence: {jax.numpy.max(kl_div)}"
 
-      max_logging.log("Checking Numerical Differences between train logits and golden logits against the provided atol, rtol.")  # pylint: disable=C0301
+      max_logging.log(
+          f"Checking Numerical Differences between train logits and golden logits against atol={test_args.rtol} rtol={test_args.atol}."
+      )
       rtol_val = float(test_args.rtol)
```
Comment (RissyRan) [inline MaxText/tests/forward_pass_logit_checker.py]: Sanity check, will atol work with kl_div if we set both? It seems so?

**Review by RissyRan — COMMENTED** [review]
Great work!

File: MaxText/tests/forward_pass_logit_checker.py (hunk 6 above)
Comment (shuningjin) [reply MaxText/tests/forward_pass_logit_checker.py]: Yes, can simultaneously check max_kl_div and atol/rtol. I add this line to make the checks explicit in log.

File: MaxText/configs/models/gpt-oss-20b.yml (hunk 2 above)
Comment (shuningjin) [reply MaxText/configs/models/gpt-oss-20b.yml]: Done.

File: MaxText/convert_gpt_oss_unscanned_ckpt.py (hunk 3 above)
Comment (shuningjin) [reply MaxText/convert_gpt_oss_unscanned_ckpt.py]: Done.

File: MaxText/convert_gpt_oss_unscanned_ckpt.py (hunk 4 above)
Comment (shuningjin) [reply MaxText/convert_gpt_oss_unscanned_ckpt.py]: Updated doc string.

File: MaxText/layers/embeddings.py (hunk 5 above)
Comment (shuningjin) [reply MaxText/layers/embeddings.py]: Updated doc string.

File: MaxText/configs/base.yml (hunk 1 above)
Comment (shuningjin) [reply MaxText/configs/base.yml]: Updated comment.

File: MaxText/configs/models/deepseek2-236b.yml [hunk 7]
```diff
@@ -49,3 +49,6 @@ original_max_position_embeddings: 4096
 rope_factor: 40
 beta_fast: 32
 mscale: 0.707
+rope_interleave: True
+rope_truncate: True
+rope_attention_scaling: False
```
Comment (RissyRan) [inline MaxText/configs/models/deepseek2-236b.yml]: nit: extra line, similar for other files

**Review by RissyRan — APPROVED** [review]
LGTM! You could get feedback first and then address conflicts before the merge.

File: MaxText/configs/models/deepseek2-236b.yml (hunk 7 above)
Comment (shuningjin) [reply MaxText/configs/models/deepseek2-236b.yml]: Done.

File: MaxText/configs/base.yml [hunk 8]
```diff
@@ -618,6 +618,9 @@ rope_factor: 40
 beta_fast: 32
 beta_slow: 1
 mscale: 1.0
+rope_interleave: True # Complex representation is interleaved or concatenated
```
Comment (gagika) [inline MaxText/configs/base.yml]: perhaps  it's more clear to have:

`# RoPE/ sin/cos interleaved vs concatenated `

File: MaxText/layers/embeddings.py
Comment (gagika) [inline MaxText/layers/embeddings.py:1]: Could you add unit test for those changes? it can be a follow up PR.

File: MaxText/configs/base.yml (hunk 8 above)
Comment (shuningjin) [reply MaxText/configs/base.yml]: Thank you! I have updated the comment.

File: MaxText/layers/embeddings.py
Comment (shuningjin) [reply MaxText/layers/embeddings.py:1]: I have added an unit test for yarn rope in `MaxText/tests/check_gpt_vs_reference.py`. It compares our updated yarn vs. huggingface gpt-oss rope.
